### PR TITLE
planner: add some test cases about partition-table dynamic mode with global-stats and SQL binding

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: related to #24150 

Problem Summary: planner: add some test cases about partition-table dynamic mode with global-stats and SQL binding

### What is changed and how it works?
No logical change, just add some test cases to test the planner can work correctly with global-stats and can choose the right index for partition tables, and SQL binding can work with dynamic-mode.

### Check List 

Tests 

- Unit test

### Release note 

- planner: add some test cases about partition-table dynamic mode with global-stats and SQL binding

### CODE REVIEW

**Review by rebelice — APPROVED** [review]

File: executor/partition_table_test.go [hunk 1]
```diff
@@ -222,6 +227,84 @@ func (s *partitionTableSuite) TestPartitionInfoDisable(c *C) {
...
+		partition p0 values in (10, 11, 12, 13, 14, 15, 16, 17, 18, 19),
+		partition p0 values in (20, 21, 22, 23, 24, 25, 26, 27, 28, 29),
+		partition p0 values in (30, 31, 32, 33, 34, 35, 36, 37, 38, 39),
+		partition p0 values in (40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50))`)
+
+	// construct some special data distribution
+	vals := make([]string, 0, 1000)
+	listVals := make([]string, 0, 1000)
+	for i := 0; i < 1000; i++ {
+		if rand.Intn(100) < 1 {
```
Comment (time-and-fate) [inline executor/partition_table_test.go:259]: Would this make it a little unstable? How about call the if branch 10 times and the else branch 990 times?

File: executor/partition_table_test.go [hunk 2]
```diff
@@ -222,6 +227,84 @@ func (s *partitionTableSuite) TestPartitionInfoDisable(c *C) {
...
+		}
+	}
+	tk.MustExec("insert into thash values " + strings.Join(vals, ","))
+	tk.MustExec("insert into trange values " + strings.Join(vals, ","))
+	tk.MustExec("insert into tlist values " + strings.Join(listVals, ","))
+
+	// before analyzing, the planner will choose TableScan to access the 1% of records
+	c.Assert(tk.HasPlan("select * from thash where a<100", "TableFullScan"), IsTrue)
+	c.Assert(tk.HasPlan("select * from trange where a<100", "TableFullScan"), IsTrue)
+	c.Assert(tk.HasPlan("select * from tlist where a<1", "TableFullScan"), IsTrue)
```
Comment (time-and-fate) [inline executor/partition_table_test.go:276]: Can we think of the plan before analyzing as a stable plan?

File: executor/partition_table_test.go (hunk 1 above)
Comment (qw4990) [reply executor/partition_table_test.go]: OK, it's updated, PTAL

File: executor/partition_table_test.go (hunk 2 above)
Comment (qw4990) [reply executor/partition_table_test.go:276]: Yes, and we have to rely on this assumption to test the impact of global-stats to our optimizer.

Comment (qw4990) [reply executor/partition_table_test.go:276]: This is a little tricky , but I cannot think out any other way to test the impact of global-stats to the optimizer.
Do you have any other idea about this? @time-and-fate

Comment (time-and-fate) [reply executor/partition_table_test.go:276]: Not really.... Actually I think the assertions after analyzing are enough.
What I'm concerning is that if there're some changes to the optimizer in the future that make the plans before analyzing change, this test might fail.

Comment (qw4990) [reply executor/partition_table_test.go:276]: Yes, this may happen, and then we should go back to this test and modify it as well. @time-and-fate

Comment (time-and-fate) [reply executor/partition_table_test.go:276]: OK.

**Pull request merged**